src/_pytest/pathlib.py

```diff
@@ -607,11 +607,20 @@ def _is_same(f1: str, f2: str) -> bool:
 
 def module_name_from_path(path: Path, root: Path) -> str:
     """
-    Return a dotted module name based on the given path, anchored on root.
+    Return a dotted module name based on the given path,
+    anchored on root or the most likely entry in `sys.path`.
 
     For example: path="projects/src/tests/test_foo.py" and root="/projects", the
     resulting module name will be "src.tests.test_foo".
     """
+    candidates = (
+        _module_name_from_path(path, dir)
+        for dir in itertools.chain([root], map(Path, sys.path))
+    )
+    return ".".join(min(candidates, key=len))  # type: ignore[arg-type]
```

Comment:
This is going to be somewhat slow I reckon. Pathlib is a hog (particularly `relative_to`) and sys.path can have many entries in some circumstances, and this is running for every import.

Reply:
I know, I just did the bare functional minimum.

I think this could be improved by having a cache that
1. stores all paths that are ever requested
2. as long as `sys.path` stays the same, otherwise the cache gets cleared